Add compatibility to searchAll and searchAny.

Ergus opened the pull request:
Related: #86605

Allow to use functions `searchAll` and `searchAny` in columns with no index and with constant strings.

With this change we can use queries like:

```sql
 SELECT searchAll('This text is a bad text', ['text', 'cat'])
 ```

and 

```sql
 SELECT count() FROM table WHERE searchAny(text, ['mytoken'])
```

and `text` doesn't need to have an index to work. It will be much slower, but may produce the same output using full scan algorithms.

### Changelog category (leave one):
- Experimental Feature

### Changelog entry (a [user-readable short description](https://github.com/ClickHouse/ClickHouse/blob/master/docs/changelog_entry_guidelines.md) of the changes that goes into CHANGELOG.md):
Functions `searchAll` and `searchAny` now work on top of columns without text columns. In those cases, they use the default tokenizer.

Ergus committed:
Add compatibility to searchAll and searchAny.

Allow to use it in columns with no index and with constant strings.

Ergus committed:
Small improvement

Ergus committed:
Update test 02346_text_index_function_search

rschu1ze reviewed (commented):
I added a reference to related issue https://github.com/ClickHouse/ClickHouse/issues/86605 to the PR message.

This PR needs docs, see the end of src/Functions/searchAnyAll.cpp and inverted_indexes.md.

Ergus committed:
Include function use in select and filter steps

Ergus committed:
FunctionSearchNeedles -> Needles

Ergus committed:
Update documentation about the changes in function behavior

Ergus commented:
> This PR needs docs, see the end of src/Functions/searchAnyAll.cpp and inverted_indexes.md.

Partially solved via 5354b55f37b3af79775808852591a71c

Ergus committed:
Simplify and improve.

Avoid using raw pointer.
Move zero match inside the execute function.
Use concepts and static_assert to detect future incomplete modification errors at compile time.
Condense logics in one responsibility principle.

rschu1ze committed:
Merge remote-tracking branch 'ClickHouse/master' into compatibility_text_search_functions

rschu1ze committed:
Update docs

Ergus committed:
Partially revert 391c136e3bc

Recover branch with `ColumnArray`
Restore token_default_extractor at class level (avoid multi-threading race condition)
Use same index to count and ordinal.

rschu1ze commented:
Known failures are all related to Iceberg and known.
